File: src/com/rian/difficultycalculator/skills/standard/StandardSpeed.java
package com.rian.difficultycalculator.skills.standard;

import com.rian.difficultycalculator.beatmap.hitobject.DifficultyHitObject;
import com.rian.difficultycalculator.evaluators.standard.StandardRhythmEvaluator;
import com.rian.difficultycalculator.evaluators.standard.StandardSpeedEvaluator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.EnumSet;

import ru.nsu.ccfit.zuev.osu.game.mods.GameMod;

/**
 * Represents the skill required to press keys or tap with regards to keeping up with the speed at which objects need to be hit.
 */
public class StandardSpeed extends StandardSkill {
    private double currentStrain;
    private double currentRhythm;
    private final ArrayList<Double> objectStrains = new ArrayList<>();
    private final double greatWindow;

    /**
     * @param mods The mods that this skill processes.
     * @param greatWindow The 300 hit window.
     */
    public StandardSpeed(EnumSet<GameMod> mods, double greatWindow) {
        super(mods);

        this.greatWindow = greatWindow;
    }

    /**
     * Calculates the number of clickable objects weighted by difficulty.
     */
    public double relevantNoteCount() {
        if (objectStrains.size() == 0)
            return 0;

        double maxStrain = Collections.max(objectStrains);

        if (maxStrain == 0)
            return 0;

        double relevantNoteCount = 0;

        for (double strain : objectStrains) {
            relevantNoteCount += 1 / (1 + Math.exp(-(strain / maxStrain * 12 - 6)));
        }

        return relevantNoteCount;
    }

    @Override
    protected double strainValueAt(DifficultyHitObject current) {
        currentStrain *= strainDecay(current.deltaTime);
        double skillMultiplier = 1375;
        currentStrain += StandardSpeedEvaluator.evaluateDifficultyOf(current, greatWindow) * skillMultiplier;
        currentRhythm = StandardRhythmEvaluator.evaluateDifficultyOf(current, greatWindow);

        double totalStrain = currentStrain * currentRhythm;

        objectStrains.add(totalStrain);

        return totalStrain;
    }

    @Override
    protected double calculateInitialStrain(double time, DifficultyHitObject current) {
        return currentStrain * currentRhythm * strainDecay(time - current.previous(0).startTime);
    }

    @Override
    protected void saveToHitObject(DifficultyHitObject current) {
        current.tapStrain = currentStrain;
        current.rhythmMultiplier = currentRhythm;
    }

    @Override
    protected double getDifficultyMultiplier() {
        return 1.04;
    }

    @Override
    protected int getReducedSectionCount() {
        return 5;
    }

    private double strainDecay(double ms) {
        return Math.pow(0.3, ms / 1000);
    }
}

File: src/com/rian/osu/difficulty/skills/StandardSpeed.kt
package com.rian.osu.difficulty.skills

import com.rian.osu.beatmap.hitobject.Slider
import com.rian.osu.difficulty.StandardDifficultyHitObject
import com.rian.osu.difficulty.evaluators.StandardRhythmEvaluator
import com.rian.osu.difficulty.evaluators.StandardSpeedEvaluator
import com.rian.osu.difficulty.utils.DifficultyCalculationUtils
import com.rian.osu.mods.Mod
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.pow

/**
 * Represents the skill required to press keys or tap with regards to keeping up with the speed at which objects need to be hit.
 */
class StandardSpeed(
    /**
     * The [Mod]s that this skill processes.
     */
    mods: Iterable<Mod>
) : HarmonicSkill<StandardDifficultyHitObject>(mods) {
    override val harmonicScale = 20.0

    private var currentDifficulty = 0.0
    private var maxDifficulty = 0.0
    private val skillMultiplier = 1.16
    private val strainDecayBase = 0.3

    private val sliderDifficulties = mutableListOf<Double>()

    /**
     * Calculates the number of clickable objects weighted by difficulty.
     */
    fun relevantNoteCount(): Double {
        if (objectDifficulties.isEmpty() || maxDifficulty == 0.0) {
            return 0.0
        }

        return objectDifficulties.sumOf { 1 / (1 + exp(-(it / maxDifficulty * 12 - 6))) }
    }

    /**
     * Obtains the amount of sliders that are considered difficult in terms of relative difficulty, weighted by consistency.
     *
     * @param difficultyValue The final difficulty value.
     */
    fun countTopWeightedSliders(difficultyValue: Double): Double {
        if (sliderDifficulties.isEmpty() || noteWeightSum == 0.0) {
            return 0.0
        }

        // What would the top note be if all note values were identical
        val consistentTopNote = difficultyValue / noteWeightSum

        if (consistentTopNote == 0.0) {
            return 0.0
        }

        // Use a weighted sum of all notes. Constants are arbitrary and give nice values
        return sliderDifficulties.sumOf {
            DifficultyCalculationUtils.logistic(it / consistentTopNote, 0.88, 10.0, 1.1)
        }
    }

    override fun objectDifficultyOf(current: StandardDifficultyHitObject): Double {
        val decay = strainDecay(current.strainTime)

        currentDifficulty *= decay
        currentDifficulty += StandardSpeedEvaluator.evaluateDifficultyOf(current) * (1 - decay) * skillMultiplier

        val currentRhythm = StandardRhythmEvaluator.evaluateDifficultyOf(current)

        val difficulty = currentDifficulty * currentRhythm
        maxDifficulty = max(maxDifficulty, difficulty)

        if (current.obj is Slider) {
            sliderDifficulties.add(difficulty)
        }

        return difficulty
    }

    private fun strainDecay(ms: Double) = strainDecayBase.pow(ms / 1000)
}
